enhancements/olm/marketplace-api-deprecation.md

```diff
@@ -0,0 +1,107 @@
+---
+title: marketplace-api-deprecation
+authors:
+  - "@ecordell"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-01-06
+last-updated: 2020-01-06
+status: implementable
+---
+
+# Marketplace Operator API Deprecation
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement covers the remaining steps needed to ensure that everything the marketplace-operator APIs have been providing is covered by other features within OLM or OpenShift.
+
+## Open Questions
+
+1. Do we wish to block an upgrade of a cluster if we can detect OperatorSource/CatalogSourceConfig usage that would be affected by the deprecation? (i.e. non-default catalogs using these features).
+2. Are there CVO limitations to be aware of? What happens if we remove a ClusterOperator instance entirely during an upgrade?
+
+## Motivation
+
+In OpenShift 4.1+, the marketplace-operator is responsible for pulling repository data into a cluster from an external source (Quay.io Appregistry) for use by OLM.
+
+As of OpenShift 4.4, alternatives for pulling external data have been built into OLM. The appregistry protocol itself is largely replaced by other efforts in OCI, and we cannot continue to rely on it long-term.
+
+### Goals
+
+- Remove support for appregistry-based storage for operator data
+- Remove the on-cluster components that support it
+- Continue to support configuration for OperatorHub
+
+### Non-Goals
+
+- Preserve non-default OperatorSource and CatalogSourceConfig data
+
+## Proposal
+
+### Removal of the Marketplace APIs
+
+Remove the OperatorSource and CatalogSourceConfig CRDs.
+
+Migrate the default OperatorSources to CatalogSources.
+
+Remove the reconcilation of the OperatorSource and CatalogSourceConfig APIs either by removing the marketplace operator entirely or by removing the relevant portions of it (if it is being retained to manage the OperatorHub API, see below).
+
+### Migration of the OperatorHub Config API
+
+The OperatorHub config API is used to configure which catalogs are available by default in a cluster, and is an important configuration point for disconnected clusters.
+
+This API is currently managed by the marketplace operator and needs to continue to be supported. This could be by continuing to support the marketplace operator as an "operatorhub config operator", by migrating the functionality into OLM, or into a new operator in the OLM repo. The decision here will also determine if we will continue to update a separate ClusterOperator object in the marketplace namespace.
```

Comment:
From an implementation perspective, it would probably be easier to leave this API as something the marketplace does and just directly interact with CatalogSources (it's probably already 99% of the way there, we would just have to modify some client calls).

But I do think that from a complexity perspective this has the potential to be relatively confusing and complicated over time. This feels a lot like technical debt that we don't want to live around forever. IMO if we don't address it *now* we will probably either never get to it or have to do it later at a much higher cost if/when this API is more complicated.

Reply:
If we managed the default catalogs as `CatalogSource` manifests with CVO directly, could we deprecate the `OperatorHub` API with a staged migration:

1. Remove existing `OperatorHub` reconciliation logic
  - Parse existing `OperatorHub` resource
  - Disable CVO reconciliation of disabled `CatalogSources`
  - Delete disabled `CatalogSources`
  - Delete `OperatorHub` resource
2. Remove `OperatorHub` CRD
  - Conditionally activate migration